noir-projects/noir-contracts/contracts/token_contract/src/main.nr

```diff
@@ -383,5 +385,63 @@ contract Token {
         storage.public_balances.at(owner).read().to_integer()
     }
     // docs:end:balance_of_public
+
+    #[aztec(private)]
+    fn split_into_partial_notes_pair(from: AztecAddress, to: AztecAddress, amount: Field, nonce: Field) -> [Field; 2] {
+        if (!from.eq(context.msg_sender())) {
+            assert_current_call_valid_authwit(&mut context, from);
+        } else {
+            assert(nonce == 0, "invalid nonce");
+        }
+
+        storage.balances.sub(from, U128::from_integer(amount));
+
+        let partial_note_hash_0 = storage.balances.add_partial(from);
+        let partial_note_hash_1 = storage.balances.add_partial(to);
+
+        let hash = compute_partial_notes_pair_hash(
+            partial_note_hash_0,
+            partial_note_hash_1,
+            context.msg_sender(),
+            amount
+        );
+
+        context.call_public_function(
+            context.this_address(),
+            FunctionSelector::from_signature("auth_partial_notes_pair((Field))"),
+            [hash.to_field()]
+        );
+
+        // emit as unencrypted for e2e testing as return values are not accessible there
+        emit_unencrypted_log_from_private(&mut context, [partial_note_hash_0, partial_note_hash_1]);
+
+        [partial_note_hash_0, partial_note_hash_1]
+    }
+
+    #[aztec(public)]
+    #[aztec(internal)]
+    fn auth_partial_notes_pair(hash: PartialNotesHash) {
+        storage.partial_notes.at(hash).write(true);
+    }
+
+    #[aztec(public)]
+    fn complete_partial_notes_pair(partial_note_hashes: [Field; 2], amounts: [Field; 2]) {
+        let hash = compute_partial_notes_pair_hash(
+            partial_note_hashes[0],
+            partial_note_hashes[1],
+            context.msg_sender(),
+            amounts[0] + amounts[1]
+        );
+
+        assert(storage.partial_notes.at(hash).read(), "Partial notes not authorized");
+        // only completable once
+        context.push_new_nullifier(hash.to_field(), 0);
```

Comment:
I actually thought this would be neater as it signals both that it was an action that was allowed at some point in the past and that it was consumed and can not be run anymore (in way, similar to contract initialiasation)

Reply:
Yeah that makes sense. I was just thinking that this uses 32 bytes more DA. Setting the storage to 0 would presumably be squashed onto the setting of it to 1 in `auth_partial_notes_pair()`